statedb v2.0 with per-table locks and delete tracking

joamaki opened the pull request:
"StateDB v2.0" reimplements pkg/statedb to introduce per-table locking, built-in revision numbers and ability to track objects being deleted from tables. It moves away from the go-memdb library and implements directly on top of the immutable-radix-tree library, with the biggest visible change being that indexes are no longer defined using reflection, but with type-safe functions.

To make reviewing easier the new version is first introduced as `pkg/statedb2` and then in a follow-up PR the uses of `pkg/statedb` are converted over to the new API and old code replaced.

See `pkg/statedb2/example` for an overview of how the new API looks like.

joestringer commented on pkg/statedb/index/keyset.go:
```diff
@@ -0,0 +1,61 @@
+
+func (ks *KeySet) Append(k []byte) {
+	if len(k) > 2<<16 {
+		panic("keyset.Append: key too long, maximum is 16k")
```
Math is off here.
```suggestion
	if len(k) >= 2<<16 {
		panic("keyset.Append: key too long, maximum is 65k")
```

joamaki replied:
Thanks fixed!

tommyp1ckles commented on pkg/statedb/example/reconcile.go:
```diff
@@ -0,0 +1,176 @@
+
+		minRevision = newRevision
+
+		if processErr != nil {
```
I think this aligns with how I've been thinking about health checking.  Only thing I'm wondering is if this is going to be a common pattern going forward where we can abstract out having to write the backoff logic/health reporting logic.

This is an example, but "backends" sound like the type of thing that might have a large amount of items being processed. I do see there's a rate limiter as well, even so I think health reporting shouldn't be subject to changing that much - I'm thinking we may need to come up with some kind of batching system for doing these types of health updates.

tommyp1ckles commented on pkg/statedb/example/reconcile.go:
```diff
@@ -0,0 +1,176 @@
+
+		if processErr != nil {
+			r.Reporter.Degraded(fmt.Sprintf("Failure at revision %d, latest is %d", minRevision, tableRevision), processErr)
+			if err := backoff.Wait(ctx); err != nil {
```
So in this case, each item has to complete processing before moving on with the next one - so this backoff would stall further reconciliation tasks?

joamaki replied:
Yep, that's the case in this example. 

I'm going to add another reconciler type which will do retrying of individual objects in the style of Resource[T]. Essentially we can do the same thing here as in Resource[T] with work queues by pushing the upsert and delete keys into the queue. Not yet sure how efficient it's going to be due to repeated creation of single-use ReadTxn and point-wise lookups. Perhaps a hybrid could work...

joamaki replied:
Ok added that (see `pkg/statedb2/workqueue.go` and `pkg/statedb2/examples/reconcile_workqueue.go`). No health status updates for them yet.

I'm going to split this and the k8s reflector into their own PRs shortly.
EDIT: currently waiting in branch `pr/joamaki/statedb2-workqueue-and-reflector`

joamaki replied:
https://github.com/joamaki/cilium/blob/pr/joamaki/statedb2-workqueue-and-reflector/pkg/statedb2/example/reconcile_workqueue.go

joamaki committed:
lock: SortableMutex for ordered locking of a mutex set

SortableMutex is a mutex paired with a (globally unique) sequence number
which allows sorting a set of them into a predicable order for ordered
locking.

This will be used in a follow-up commit to implement per-table locks for
StateDB in a way that avoids deadlocks due to misordered locking sequence
(A->B and B->A).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb2: StateDB v2.0 with per-table locks and deletion tracking

This introduces an evolution of pkg/statedb that solves the following
issues:
  * Per-table locking instead of global writer lock for increased resiliency.
  * No more use of reflection in indexing. Indexes are defined in a type-safe
    way.
  * Revisions are built into the database so it does not need to be reimplemented
    by each table.
  * Deletions on a table can now be tracked and a consistent sequence of inserts
    and deletes can be derived from it to e.g. implement retryable reconciliation
    loops.

It is introduced as pkg/statedb2 for easier code reviewing (no need to rewrite
uses of pkg/statedb in this PR) and to allow some time for it to mature before
replacing pkg/statedb.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb2: Add example application

Add an example usage of statedb2 that showcases the pattern for
implementing a resilient reconciliation loop on top of the library.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb2: Simple fuzz test

Add a simple fuzz test to test different random combinations of write transactions
and operations.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/lock/sortable_mutex.go:
```diff
@@ -0,0 +1,64 @@
+// SortableMutex provides a Mutex that can be globally sorted with other
+// sortable mutexes. This allows deadlock-safe locking of a set of mutexes
+// as it guarantees consistent lock ordering.
+type SortableMutex struct {
```
Since that `sync.Mutex`es are usable by default, while this isn't, I think the struct should not be exported.

joamaki replied:
Ah good catch. I hid the struct behind an inteface, so it cannot now be initialized wrong.

bimmlerd commented on pkg/lock/sortable_mutex.go:
```diff
@@ -0,0 +1,64 @@
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s SortableMutexes) Lock() {
```
I think this deserves a comment to the effect of:

`Lock sorts the mutexes, and then locks them in order. If any lock cannot be acquired, this will block while holding the locks with a lower sequence number.`

joamaki replied:
Thanks, added!

bimmlerd commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,177 @@
+// it until Commit() is called. To discard the changes call Abort().
+//
+// WriteTxn is not thread-safe!
+func (db *DB) WriteTxn(table TableMeta, tables ...TableMeta) WriteTxn {
```
I think we'll need to be pretty careful about deadlocks when acquiring write transactions. Clearly, the total lock order of the tables prevents deadlocks between tables, but it's still possible for table locks to be included in a wait cycle. Here's an example:
Let's call the ipcache mutex `Mipcache`, and some table Ta's lock `MTa`. If any code path acquires `Mipcache` and then a write transaction including `Ta` (i.e. locking `MTa`), while any other path acquires `Mipcache` while holding `MTa`, we have a deadlock.
It may be prudent to prohibit acquiring _any_ external locks while holding a `WriteTxn`. (If not enforceable via some custom linter, we should at least document it).

joamaki replied:
Yep that's a good idea to document! For somewhat simple cases that don't involve function pointers or interfaces we can probably catch this with a linter.

dylandreimerink committed:
statedb2: Add metrics

This commit adds metrics to statedb2, metrics mostly center around
locking, contention, number of entries and duration of certain
operations since these topics are the most interesting to us.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,219 @@
+type DB struct {
+	tables    map[TableName]TableMeta
+	mu        lock.Mutex // sequences modifications to the root tree
+	root      atomic.Pointer[iradix.Tree[tableEntry]]
```
Lookup performance can be significantly improved by switching to a flat array presentation for the tables and indexes, e.g. assign each table a position, each index a position within the table and store those positions within `genTable` and `Index`:
```
type DB struct {
  ...
  rootPtr atomic.Pointer[root]
  ... 
}

type root []tableEntry

type tableEntry struct {
  indexes []iradix.Tree[object]
  revision Revision
}

tables := db.root.Load()
index := tables[tablePos].indexes[indexPos]
```

This would be especially beneficial for "point-wise lookups" (`table.First(db.ReadTxn(), index.Query(key))`) that currently go through 3 radix trees. I'll leave this optimization for a follow-up PR for now though.

joamaki commented on pkg/statedb2/metrics.go:
```diff
@@ -0,0 +1,89 @@
+func NewMetrics() Metrics {
+	return Metrics{
+		WriteTxnDuration: metric.NewHistogramVec(metric.HistogramOpts{
+			Namespace: metrics.CiliumAgentNamespace,
```
The namespace should be configurable via `NewDB`... or via metrics registry?

dylandreimerink replied:
The registry should set the namespace of very collected metric.

bimmlerd commented on pkg/statedb2/cell.go:
```diff
@@ -0,0 +1,57 @@
+//
+// For example use see pkg/statedb/example.
+var Cell = cell.Module(
+	"statedb",
```
this should probably differentiate it-self more clearly from statedb

joamaki replied:
Right yeah, since they'll live side-by-side for a few days, I'll rename this to statedb2.

bimmlerd commented on pkg/statedb2/index/int.go:
```diff
@@ -0,0 +1,18 @@
+
+func Uint64(n uint64) Key {
+	buf := make([]byte, 8)
+	binary.BigEndian.PutUint64(buf, n)
```
A comment explaining why big endian lends itself better to this task would be nice

bimmlerd commented on pkg/statedb2/index/keyset_test.go:
```diff
@@ -0,0 +1,53 @@
+	ks.Append([]byte("foo"))
+	require.EqualValues(t, "foo", ks.First())
+	ks.Foreach(func(bs []byte) {
+		require.EqualValues(t, "foo", ks.First())
```
I think this should compare `bs`

bimmlerd commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+var zeroTxn = txn{}
+
+func revisionKey(rev uint64, idKey []byte) []byte {
+	buf := make([]byte, 8+len(idKey))
```
putting the 8 in a named constant would help with quicker understanding

bimmlerd commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+	if indexTreeTxn, ok := txn.writeTxns[tableIndex{name, index}]; ok {
+		return indexTreeTxn
+	}
+	table, ok := txn.tables[name]
```
there's a couple of these panics based on not finding the table; I think there's `getTable` for this already

bimmlerd commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+			panic("BUG: Table " + tableIndex.table + " not cached")
+		}
+		table.indexes, _, _ =
+			table.indexes.Insert([]byte(tableIndex.index), subTxn.CommitOnly())
```
you're explicitly ignoring the error here. I think it's non-obvious why this can't fail, so maybe an explanatory comment would help? :)

joamaki replied:
What error? Insert cannot fail, it returns `(newTree Tree, oldObject T, oldExisted bool)`

bimmlerd commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+
+	// Insert the modified tables into the root.
+	for name, table := range txn.tables {
+		rootTxn.Insert([]byte(name), *table)
```
remark on consistency: I think the following properties hold:

1. the critical section of updating the root is protected from concurrency by `db.mu`
2. Concurrent write transactions can only occur if the set of tables the write transactions cover is disjoint (thanks to the sortable mutexes), and hence there cannot be a conflict of writers here.
3. Read transactions can occur concurrently to a write transaction (even with overlapping tables), but will observe a consistent state (as long as the immutability of objects within the radix is upheld).

Something like this somewhere in the code would help; having the properties formulated makes it easier to think about how they can be broken

bimmlerd commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+	}
+	rootTxn.Notify()
+
+	txn.smus.Unlock()
```
do we need to keep the table locks before we notify? (I think this came up in the code walk too)

joamaki replied:
We don't, we can drop them and then notify.

bimmlerd commented on pkg/statedb2/index/keyset.go:
```diff
@@ -0,0 +1,62 @@
+	}
+}
+
+func (ks KeySet) Exists(k Key) bool {
```
I think the bit banging is correct, but in general the keyset deserves some comments, maybe even some ascii art :innocent:

bimmlerd commented on pkg/statedb2/db.go:
```diff
@@ -90,9 +90,7 @@ type DB struct {
-		tables:    make(map[TableName]TableMeta),
-		gcTrigger: make(chan struct{}, 1),
-		gcExited:  make(chan struct{}),
+		tables: make(map[TableName]TableMeta),
```
these changes might want to fixedup into the initial commit?

bimmlerd commented on pkg/lock/sortable_mutex.go:
```diff
@@ -16,18 +17,28 @@ var sortableMutexSeq atomic.Uint64
 type SortableMutex interface {
 	sync.Locker
 	Seq() uint64
+	AcquireDuration() time.Duration // The amount of time it took to acquire the lock
```
maybe `cumulative amount of time` in this case?

bimmlerd reviewed (approved):
it's a lot of code to read, so couldn't review it all thoroughly- mostly skipped the metrics, don't know much about those anyway.

Other than that, impressive feat of engineering.

tommyp1ckles commented on pkg/lock/sortable_mutex.go:
```diff
@@ -0,0 +1,76 @@
+// sortableMutex implements SortableMutex. Not exported as the only way to
+// initialize it is via NewSortableMutex().
+type sortableMutex struct {
+	*sync.Mutex
```
we're not going to use our internal mutex type here?

joamaki replied:
Ah good catch. I'll switch it to `lock.Mutex`.

tommyp1ckles commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,369 @@
+}
+
+func (txn *txn) Insert(meta TableMeta, data any) (any, bool, error) {
+	if txn.rootReadTxn == nil {
```
Where does this ever get set to nil?

joamaki replied:
Abort and Commit set it to mark the txn as finished. I'll think about ways to clean this up a bit.

tommyp1ckles commented on pkg/statedb2/table.go:
```diff
@@ -0,0 +1,214 @@
+	return &iterator[Obj]{iter}, watchCh
+}
+
+// Insert implements Table
```
unfinished comment?

joamaki replied:
Ah this was auto-filled by golsp. I'll clean these up.

tommyp1ckles commented on pkg/statedb2/index/keyset.go:
```diff
@@ -0,0 +1,62 @@
+
+// KeySet is a sequence of (length, byte slice) pairs.
+// length is encoded as 16-bit big-endian unsigned int.
+type KeySet struct {
```
why does this need to be in a specific byte order instead of just an int type?

joamaki replied:
sure, I might as well make it native 32-bit int and use `binary.NativeEndian`. I'm still wondering though if I'm prematurely optimizing here and `[][]byte` would be just fine. I wanted to keep memory allocations to somewhat minimum on inserts...

tommyp1ckles commented on pkg/statedb2/table.go:
```diff
@@ -0,0 +1,214 @@
+
+func NewTable[Obj any](
+	tableName TableName,
+	primaryIndexer Indexer[Obj],
```
is it obvious that the primary indexer only ever use the first key value in the set?

joamaki replied:
Primary index must not be a multi-index which is why only first key is used. I'll think about whether it'd be worth the extra code to split into SingleIndexer and MultiIndexer to make this explicit or see if I can figure out some other ways, perhaps allowing specifying either `FromObject func(Obj) index.Key` or `FromObjectMulti func(Obj) index.KeySet`, but not both.

tommyp1ckles replied:
documenting this behaviour might be sufficient

tommyp1ckles reviewed (commented):
Few questions - might take me another pass or two before I understand this enough to make meaningful comments.

dylandreimerink commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,233 @@
+
+	db.metrics.WriteTxnAcquisition.With(prometheus.Labels{
+		"package": callerPkg,
+		"tables":  strings.Join(tableNames, "+"),
```
We (I) forgot to sort the table names. If we don't do that then metrics for write transactions for `table1, table2` and `table2, table1` will end up in separate buckets.

dylandreimerink commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,233 @@
+	}
+}
+
+func (db *DB) Start(hive.HookContext) error {
+	db.gcTrigger = make(chan struct{}, 1)
+	db.gcExited = make(chan struct{})
+	go graveyardWorker(db)
+	return nil
+}
+
+func (db *DB) Stop(ctx hive.HookContext) error {
+	close(db.gcTrigger)
+	select {
+	case <-ctx.Done():
+		return errors.New("timed out waiting for graveyard worker to exit")
+	case <-db.gcExited:
+	}
+	return nil
+}
```
nit: not strictly necessary, but we could use a timer job + trigger here instead of managing the lifecycle of this goroutine manually.

dylandreimerink commented on pkg/statedb2/graveyard.go:
```diff
@@ -0,0 +1,130 @@
+		txn = db.WriteTxn(tablesToModify[0], tablesToModify[1:]...).getTxn()
+		for meta, deadObjs := range toBeDeleted {
+			tableName := meta.Name()
+			cleaningTimes[tableName].Start()
```
I made another mistake here :disappointed: . I thought that running `.Start` on a timestat would accumulate, but it just resets the start timer. So this doesn't work as intended, we have to add a second set of timestats and combine them when reporting the CleaningDuration below

dylandreimerink commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,400 @@
+		return 0, fmt.Errorf("table %q not locked for writing", tableName)
+	}
+	table.revision++
+	txn.db.metrics.TableRevision.With(prometheus.Labels{
```
Uhm :thinking:, if we `Abort` this transaction, is the revision still incremented? Or are changes to the revision also discarded? If they are we should also switch to tracing a revision delta and only applying to the metric once we `Commit`.

dylandreimerink commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,400 @@
+	rootTxn.Notify()
+
+	txn.smus.Unlock()
+	for name, delta := range txn.pendingObjectDeltas {
```
We proccess the pending object deltas here, as we should, but I forgot to add the same logic for the `pendingGraveyardDeltas` :face_exhaling:

dylandreimerink commented on pkg/statedb2/txn.go:
```diff
@@ -0,0 +1,400 @@
+		_, obj, ok := iter.Next()
+		for ok {
+			buf.WriteString("    ")
+			bs, err := json.Marshal(obj.data)
```
If we are going to use reflection based Marshalling anyway, way do the whole manual JSON crafting instead of using it for the whole JSON message? Is this premature optimization, and if so, would another JSON library not be a better idea?

dylandreimerink reviewed (dismissed):
Nice work, though I spotted a number of mistakes I made myself in the metrics part

giorio94 commented on pkg/lock/sortable_mutex.go:
```diff
@@ -16,18 +17,28 @@ var sortableMutexSeq atomic.Uint64
 type sortableMutex struct {
-	*sync.Mutex
-	seq uint64
+	sync.Mutex
+	seq             uint64
+	acquireDuration time.Duration
```
Nit. I guess that most (all?) changes in this file should belong to ff6d098eb54774db75288680be49f29675fd0562.

giorio94 commented on pkg/statedb2/types.go:
```diff
@@ -0,0 +1,219 @@
+	// iteration of updates and deletions (see (*DeleteTracker[Obj]).Process).
+	Delete(WriteTxn, Obj) (oldObj Obj, hadOld bool, err error)
+
+	DeleteAll(WriteTxn) error
```
Nit. The comment for `DeleteAll` is missing.

giorio94 commented on pkg/statedb2/types.go:
```diff
@@ -0,0 +1,219 @@
+
+// Iterator for iterating objects returned from queries.
+type Iterator[Obj any] interface {
+	// Next returns the next object its revision if ok is true, otherwise
```
Nit.

```diff
- // Next returns the next object its revision if ok is true, otherwise
+ // Next returns the next object and its revision if ok is true, otherwise
```

giorio94 commented on pkg/statedb2/types.go:
```diff
@@ -0,0 +1,219 @@
+
+var _ Indexer[struct{}] = &Index[struct{}, bool]{}
+
+//nolint:unused
```
Nit. I'd suggest to expand the comment to mention that nolint is required because the linter doesn't understand that these methods are used to implement a generic interface.

giorio94 commented on pkg/statedb2/types.go:
```diff
@@ -0,0 +1,219 @@
+
+	// All returns an iterator for all objects in the table and a watch
+	// channel that is closed when the table changes.
+	All(ReadTxn) (Iterator[Obj], <-chan struct{})
```
Nit. I wonder if it could make sense to define a dedicated alias type instead of using `<-chan struct{}` to signal that a change has occurred.

giorio94 commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,233 @@
+// ReadTxn constructs a new read transaction for performing reads against
+// a snapshot of the database.
+//
+// ReadTxn is not thread-safe!
```
I think that it would be worth expanding the comment here (and for WriteTxn) about thread safeness. For instance, in the example there are two `WriteTxn` operations that could possibly occur in parallel, which seems to contradict what's written here.

giorio94 commented on pkg/statedb2/deletetracker.go:
```diff
@@ -0,0 +1,111 @@
+// is returned by the function the iteration is stopped and the revision at which
+// processing failed and the error is returned. The caller can then retry processing
+// again from this revision by providing it as the 'minRevision'.
+func (dt *DeleteTracker[Obj]) Process(txn ReadTxn, minRevision Revision, processFn func(obj Obj, deleted bool, rev Revision) error) (Revision, <-chan struct{}, error) {
```
Nit. It seems a bit strange to me that both updates and deletes are watched through a `DeleteTracker`. I wonder if it could make sense to have a dedicated interface with a more semantic name for that.

giorio94 reviewed (commented):
Looks really cool! I still have to fully digest all the logic, but I left a few minor comments inline.

tommyp1ckles commented on pkg/statedb2/example/reconcile.go:
```diff
@@ -0,0 +1,176 @@
+		}
+
+		// Refresh the read transaction to read the new version of the state.
+		txn = r.DB.ReadTxn()
```
Do we need to close the old transaction here?

dylandreimerink replied:
Read transactions are like snapshots, they do not need to be closed. Only write transactions need to be Committed or Aborted.

tommyp1ckles commented on pkg/statedb2/metrics.go:
```diff
@@ -0,0 +1,89 @@
+		TableContention: metric.NewGaugeVec(metric.GaugeOpts{
+			Namespace: metrics.CiliumAgentNamespace,
+			Subsystem: "statedb",
+			Name:      "table_contention_seconds",
```
Should this be `table_contention_duration_seconds`, similar to other metrics?

tommyp1ckles commented on pkg/statedb2/metrics.go:
```diff
@@ -0,0 +1,89 @@
+	// How long it took to acquire a write transaction for all tables.
+	WriteTxnAcquisition metric.Vec[metric.Observer]
+	// How long writers were blocked while waiting to acquire a write transaction for a specific table.
+	TableContention metric.Vec[metric.Gauge]
```
These are basically measuring db lock acquire latency, should we be using histograms for this?

dylandreimerink replied:
We already have a histogram above it which measures how long it takes for all locks in a transaction to be acquired, which should cover most information. The table contention metric is just slightly more specific since it is per table, using a gauge was an attempt to keep the amount of metric data down. This is just 1 value per table, histograms have lots of values per label.

tommyp1ckles commented on pkg/statedb2/metrics.go:
```diff
@@ -0,0 +1,89 @@
+	TableGraveyardCleaningDuration metric.Vec[metric.Observer]
+}
+
+func NewMetrics() Metrics {
```
These metrics being all labelled by table (as well as possibly by package) would seem to have a pretty high metrics cardinality (up to: nodes X tables X package). 

Should these be enabled by default? and perhaps we should have a way to emit these on aggregate without tables for more general usage (otherwise these seem like they may get disabled and be largely used for developer debugging purposes)

dylandreimerink replied:
Yea, I see what you are saying, its a concern. A lot of these metrics are to track things that we as developers are worried about. The idea being that if something goes wrong and we get a sysdump that we can hopefully see if certain things are going wrong in statedb. 

But metrics might not be the correct tool for the job, or at least not in this way. Perhaps we should look into a setting where metrics are part of the CLI output and thus the sysdump and the prometheus server.

tommyp1ckles replied:
I think these metrics may be potentially useful to some users.  It may also help us reveal scaling issues in the future.

This is kind of part of a broader problem we have that I've been thinking about.  We find ourselves in a position where we walk a line between offering metrics that are high cardinality and providing ones that can be enabled by default, or we make compromises (i.e. see map_pressure metric).

Not sure what the answer is here, I think in this case we may just want to think about what we enable by default.

dylandreimerink replied:
Would love your feedback on https://github.com/cilium/cilium/issues/27659

tommyp1ckles commented on pkg/statedb2/metrics.go:
```diff
@@ -0,0 +1,89 @@
+			Subsystem: "statedb",
+			Name:      "write_txn_duration_seconds",
+			Help:      "How long a write transaction was held.",
+		}, []string{"tables", "package"}),
```
Having a package label seems especially a bit worrying since there's a large set of values this could be.

dylandreimerink replied:
The idea is that this can help identify a culprit if we ever run into high table contention situations. But using metrics as a debugging tool like this might not be the right way to go about it.

tommyp1ckles replied:
I think these might be nice to have, but maybe not enabled by default... which ultimately may make it hard to catch high table contention situations.

dylandreimerink replied:
Addressed part of the concern in https://github.com/cilium/cilium/pull/27657

tommyp1ckles commented on pkg/statedb2/db.go:
```diff
@@ -0,0 +1,233 @@
+//
+//	      T(bar).indexes
+//		   /  \
+//		  i    I(byRevision)
```
I get why we store tables in the trie, why do we need to store indices in one as well?

dylandreimerink replied:
We store the actual data in immutable radix tries which allow us to do read snapshots. But since we have multiple indecies and multiple tables which should all be part of the same read snapshot, we place the smaller tries in larger tries, until we have 1 single root node for the whole StateDB so we can get a transaction with a single atomic pointer read and update with a single atomic pointer write.

joamaki replied:
I also said in an earlier comment that this can be optimized a lot by just using arrays for tables and indexes. I will follow up with that in a later PR.
